File: app/models/instrument.py
from sqlalchemy import Column, String, Enum, Boolean, Float, DateTime
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.db.database import Base
import enum
import uuid

class InstrumentType(str, enum.Enum):
    CRYPTO = "crypto"
    STOCK = "stock"

class Instrument(Base):
    __tablename__ = "instruments"
    
    instrument_id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    symbol = Column(String(20), nullable=False, unique=True, index=True)  # e.g., "BTC", "AAPL"
    name = Column(String(100), nullable=False)  # e.g., "Bitcoin", "Apple Inc."
    type = Column(Enum(InstrumentType), nullable=False, index=True)
    current_price = Column(Float, nullable=False, default=0.0)
    price_change_24h = Column(Float, nullable=True)  # 24h price change in percentage
    volume_24h = Column(Float, nullable=True)
    market_cap = Column(Float, nullable=True)
    is_active = Column(Boolean, default=True, index=True)
    last_updated = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # Relationships
    orders = relationship("Order", back_populates="instrument")
    positions = relationship("Position", back_populates="instrument")
    trades = relationship("Trade", back_populates="instrument")
    
    def __repr__(self):
        return f"<Instrument(symbol={self.symbol}, name={self.name}, type={self.type})>"

File: frontend/trading_ecosystem/lib/models/instrument.dart
class Instrument {
  final String instrumentId;
  final String symbol;
  final String name;
  final InstrumentType type;
  final double currentPrice;
  final double? priceChange24h;
  final double? volume24h;
  final double? marketCap;
  final bool isActive;
  final DateTime lastUpdated;
  final String? tvSymbol; // TradingView symbol (e.g., "BINANCE:BTCUSDT")

  Instrument({
    required this.instrumentId,
    required this.symbol,
    required this.name,
    required this.type,
    required this.currentPrice,
    this.priceChange24h,
    this.volume24h,
    this.marketCap,
    required this.isActive,
    required this.lastUpdated,
    this.tvSymbol,
  });

  factory Instrument.fromJson(Map<String, dynamic> json) {
    return Instrument(
      instrumentId: json['instrument_id'],
      symbol: json['symbol'],
      name: json['name'],
      type: InstrumentType.values.firstWhere(
        (e) => e.value == json['type'],
        orElse: () => InstrumentType.CRYPTO,
      ),
      currentPrice: (json['current_price'] as num).toDouble(),
      priceChange24h: json['price_change_24h'] != null 
          ? (json['price_change_24h'] as num).toDouble() 
          : null,
      volume24h: json['volume_24h'] != null 
          ? (json['volume_24h'] as num).toDouble() 
          : null,
      marketCap: json['market_cap'] != null 
          ? (json['market_cap'] as num).toDouble() 
          : null,
      isActive: json['is_active'] ?? true,
      lastUpdated: DateTime.parse(json['last_updated']),
      tvSymbol: json['tv_symbol'],
    );
  }

  factory Instrument.fromMarketData(Map<String, dynamic> json) {
    return Instrument(
      instrumentId: json['symbol'] ?? json['instrument_id'] ?? '',
      symbol: json['symbol'] ?? '',
      name: json['name'] ?? '',
      type: InstrumentType.values.firstWhere(
        (e) => e.value == json['type'],
        orElse: () => InstrumentType.CRYPTO,
      ),
      currentPrice: (json['current_price'] as num?)?.toDouble() ?? 0.0,
      priceChange24h: json['price_change_24h'] != null 
          ? (json['price_change_24h'] as num).toDouble() 
          : null,
      volume24h: json['volume'] != null 
          ? (json['volume'] as num).toDouble() 
          : null,
      marketCap: json['market_cap'] != null 
          ? (json['market_cap'] as num).toDouble() 
          : null,
      isActive: true,
      lastUpdated: json['last_updated'] != null 
          ? DateTime.parse(json['last_updated'])
          : DateTime.now(),
      tvSymbol: json['tv_symbol'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'instrument_id': instrumentId,
      'symbol': symbol,
      'name': name,
      'type': type.value,
      'current_price': currentPrice,
      'price_change_24h': priceChange24h,
      'volume_24h': volume24h,
      'market_cap': marketCap,
      'is_active': isActive,
      'last_updated': lastUpdated.toIso8601String(),
      'tv_symbol': tvSymbol,
    };
  }

  @override
  String toString() {
    return 'Instrument(symbol: $symbol, name: $name, type: $type, price: \$${currentPrice.toStringAsFixed(2)})';
  }
}

enum InstrumentType {
  CRYPTO('crypto'),
  STOCK('stock');

  const InstrumentType(this.value);
  final String value;
}
